### ISSUE TITLE
[Glide Feature Request] Option to use nGlide wrapper?

### ISSUE THREAD

**Issue by Frank-74** [project64/project64#1152]
Would it be an easy thing to do?

An option in Glide64 to use an installed nGlide wrapper, instead of the bundled wrapper.

http://www.zeus-software.com/downloads/nglide

**Comment by AmbientMalice** [comment]
Out of curiosity, why would you use nglide instead?

**Comment by LegendOfDragoon** [comment]
>Would it be an easy thing to do?

nGlide is closed source and PJ64 statically links the current wrapper. So i wouldn't call it easy to do.

**Comment by Frank-74** [comment]
I can run Glide64 Final and the nGlide wrapper with maximum resolution fullscreen.  Full speed, rock steady 60V/I's.  Its much faster than the old glide3x.dll wrapper for me.

By the way, just noticed [backspace] no longer works to change the filtering mode in PJ64Glide.

**Comment by LegendOfDragoon** [comment]
Tbh, I never got nglide working ;/ . It would just give me a black screen :disappointed: .

It certainly would be cool to see Glide64 sped up. Its performance is atrocious in 2D games.

**Comment by AmbientMalice** [comment]
>By the way, just noticed [backspace] no longer works to change the filtering mode in PJ64Glide.

This is deliberate. The filtering mode is usually set for a reason, and people keep hitting BACKSPACE and changing it, and then don't know which is the "correct" mode.

**Comment by Frank-74** [comment]
It only works fullscreen.  You have to Alt+Enter to fullscreen.  Use the nGlide configurator first.

It converts all the calls to D3D instead of OGL.

**Comment by Frank-74** [comment]
This is interesting, from the nGlide readme.
Direct3D reference rasterizer (software rendering) (place d3dref9.dll from DirectX SDK into game directory to use it)

I'll give that a try tomorrow.

**Comment by Frank-74** [comment]
Can't get the Direct3D software rendering to work.

```2016/07/10 15:04:32.870 03096,Error,Exception Handler,CMipsMemoryVM::FilterX86Exception,Invalid memory adderess: AD450000
2016/07/10 15:04:32.871 03096,Error,RSP,CN64System::RunRSP,exception generated
2016/07/10 15:04:32.871 03096,Error,User Interface,CNotificationImp::FatalError,CN64System::RunRSP()
Unknown memory action

Emulation stop
```

**Comment by project64** [comment]
the opengl es code looks like it is more performance orientated, could look at using https://chromium.googlesource.com/angle/angle/+/master/README.md to get a d3d version.

**Comment by Frank-74** [comment]
@project64 
A d3d version of Glide64 would be an advantage to anyone with rubbish onboard Intel/Ati graphics.  That have rubbish OpenGL support.

D3D is a lot faster on cheap gear.

**Comment by LegendOfDragoon** [comment]
>A d3d version of Glide64 would be an advantage to anyone with rubbish onboard Intel/Ati graphics.

Even decent ATI cards have issues with certain OpenGL software. I think part of the problem is that people will default to blaming drivers, when sometimes the problem is the source code itself.

**Comment by cxd4** [comment]
>D3D is a lot faster on cheap gear.

Also constituent of the network effect.

D3D was better-documented and more profitable for industries to spend more of their time making proprietary Windows drivers for, with better performance and quality.

>I think part of the problem is that people will default to blaming drivers, when sometimes the problem is the source code itself.

IDK.  It's kinda both really.  Usually if the OpenGL drivers have a problem we're lucky to be able to code a workaround in the code itself.  It's sometimes a hack, sometimes an improvement, sometimes both or neither, but if the original code should have worked but didn't it technically is due to non-conforming drivers.

>the opengl es code looks like it is more performance orientated, could look at using https://chromium.googlesource.com/angle/angle/+/master/README.md to get a d3d version.

This would be better than rewriting Glide64 against Direct3D, because that can still use Direct3D.

**Comment by project64** [comment]
having the angel library would mean on windows it was d3d, android opengl es, but not sure what that means for the linux version.

**Comment by LegendOfDragoon** [comment]
>IDK. It's kinda both really. Usually if the OpenGL drivers have a problem we're lucky to be able to code a workaround in the code itself. It's sometimes a hack, sometimes an improvement, sometimes both or neither, but if the original code should have worked but didn't it technically is due to non-conforming drivers.

I agree. I feel that some issues probably are driver related, such as poor bliting performance and black screen. Some other issues are very questionable, however. For instance with Glide64 Final and PJ64's Glide, i have this issue in SSB where kirby's face is dark. It's not dark when using Mesa 3D. Also, libretro's Glide2Gl does not have this issue.
![image](https://cloud.githubusercontent.com/assets/10492645/16718066/e77efcde-46d0-11e6-96fd-d1aeb4cf3a3f.png)

People like to assume the bad performance I get in Glide64 and z64gl are due to drivers / hardware, but I'm convinced it's not the real reason. In general, I get good performance in glN64 and Rice 6.1 OGL, so I think the problem is the source code itself.

**Comment by project64** [comment]
there is two copies of the code in the glitch library one for OGL one for OGL ES, the ES version looks like it is likely to be more optimized, I would like to see if I could get this merged, or maybe I look at dropping the OGL version and use angel that I mentioned above and see what the performance is like, but anyway I would dislike having two copies of the same code. But for me this is after the next version is released.

**Comment by cxd4** [comment]
>having the angel library would mean on windows it was d3d, android opengl es, but not sure what that means for the linux version.

The only possibility is desktop OpenGL.  This is the way WebGL works under JavaScript:  as you said, d3d on windows, GL ES on mobile phones and plain GL on most other Unix-based systems.

>For instance with Glide64 Final and PJ64's Glide, i have this issue in SSB where kirby's face is dark. It's not dark when using Mesa 3D. Also, libretro's Glide2Gl does not have this issue.

If changing to another driver like Mesa3D fixes the issue, this usually induces the driver as the source of the problem, but that doesn't mean the GL coding isn't deficient, at fault, or can do nothing about the issue either.

>there is two copies of the code in the glitch library one for OGL one for OGL ES, the ES version looks like it is likely to be more optimized, I would like to see if I could get this merged, or maybe I look at dropping the OGL version and use angel that I mentioned above and see what the performance is like, but anyway I would dislike having two copies of the same code.

GL ES specifications **forces** programs to be faster than some desktop GL programs.
Desktop GL specifications **allows** some programs to be faster than some GL ES programs.

Basically GL ES is a subset of not only the performance optimizations you could do with regular desktop OpenGL, but a subset also of the slower, primitive and not-as-fast things that you can do with desktop OpenGL.  So it's a double-edged sword but less relevant in that ES is all that pertains to the Android port anyway.

Overall I would agree with dropping the OGL version and try using ES on Android and even Windows if you can.

**Comment by Frank-74** [comment]
Maybe a new branch experiment with Angel would be useful to compare.

D3D on Windows has allways been faster than OGL.

Would this be a 2.4 or 2.5 milestone?

**Comment by project64** [comment]
not sure, need to get 2.3 out and see what needs to be done for 2.4 .. but I would be interested in doing it but it does cause issues for @cxd4  linux effort as it most likely. Tho looks like like https://sourceforge.net/projects/dgles/ might help there. I would like to see it in 2.4 but might be 2.5. Tho If I use angel and make it d3d, do you think I should rename it to not cause confusion?

**Comment by LegendOfDragoon** [comment]
>If changing to another driver like Mesa3D fixes the issue, this usually induces the driver as the source of the problem, but that doesn't mean the GL coding isn't deficient, at fault, or can do nothing about the issue either.

I'm kinda 50 / 50 on that. For common problems like black screen, I'm leaning towards driver issue. But I'm not so sure about the plugin specific issues. I may try and see what code makes the difference between Glide2GL and PJ64's Glide64.

>Tho If I use angel and make it d3d, do you think I should rename it to not cause confusion?

Yeah, renaming would help prevent confusion.

**Comment by Frank-74** [comment]
Anyway, back on topic.  It's not possible to provide an option to use an alternate glitch (glide3x) library, as glitch is statically linked?

Another question.  Has Glide64 for PJ64 had fixes/enhancements that improve performance, on top of Glide64 Final?

I ask this because of the menu's in Monaco Grand Prix are fast to display in Glide64 for PJ64, but in Final, they take about a second per level of shade on a fade in/out to the next screen.  Moving to select another option is horibbly slow in Final, but acceptable in PJGlide64.

Cheers

**Comment by cxd4** [comment]
>but I would be interested in doing it but it does cause issues for @cxd4 linux effort as it most likely.

The difficulty in getting GL ES level commands supported on your Windows port of Project64 Glide64 is greater than the difficulty of getting it to work on Linux.

>For common problems like black screen, I'm leaning towards driver issue. But I'm not so sure about the plugin specific issues.

Oh well if it's plugin-specific, then it could be either one, yeah.

Texture crashing and such could sometimes be inferred from the fact that system memory (set up client-side from C) may have been set from the result of bugs before even reaching the OpenGL server, so technically it could also be neither a driver issue nor an OpenGL coding issue.

>Anyway, back on topic. It's not possible to provide an option to use an alternate glitch (glide3x) library, as glitch is statically linked?

Code-wise it's still in the stage where it could be processed back as an external DLL instead of statically linking the code into one binary plugin.  (Some find this disadvantageous because this requires the legacy GLIDE call names in the plugin's source instead of directly writing out the GL commands intended for use.)

So maybe it could be compiled back into a glide3x.dll externally linked thing, but I see no advantage to doing this just so you can use nGlide.  nGlide is some Direct3D thing that we really don't need, and the new shader-based Glitch64 wrapper is much faster than the glide3x version you've tested with Glide64 Final for speed.  Up-to-date OpenGL coding is faster than Direct3D, even on Windows.  That and OpenGL has no native C++ language binding element like DirectX COM does.  Not like having Windows operating system installed in the first place helps performance anyway.

**Comment by LegendOfDragoon** [comment]
>@LegendOfDragoon It only works fullscreen. You have to Alt+Enter to fullscreen. Use the nGlide configurator first.

Thanks a lot for posting this! I've tried nglide before but dismissed it, since I just got a black screen lol.. It's a shame it only works in fullscreen, but my gaming experience is way better with nglide than any other wrapper I've tried! I don't even think there's a reason for me to use Jabo's anymore, other than occasional testing. With nglide, I no longer get software rendering tier performance. To think changing wrappers would make this much of a difference.. I honestly thought nglide was some meme, but it's nothing short of amazing.

If only nglide was open source... Anyway, I'm glad to know that the issue has been narrowed down. The wrapper itself is what needs some serious work.

>Another question. Has Glide64 for PJ64 had fixes/enhancements that improve performance, on top of Glide64 Final?

Tbh, I haven't really seen an instance where PJ64's version was significantly faster.

**Comment by Frank-74** [comment]
@LegendOfDragoon 
Exiting fullscreen is a bit dodgy with nGlide.  I've found a way that doesn't make Project64 lockup.

Use only Alt+Enter to enter fullscreen.  If the PJ64 menu shows up, press Alt+Enter again.
To exit press Escape and then F12 to End Emulation and go back to the ROM browser.

In the nGlide history of updates, way way back.  nGlide was made compatible with Project 64.  But this must have been for 1.6.  I suppose a bug report for using with PJ64 2.3+ might help them fix this issue.

**Comment by LegendOfDragoon** [comment]
Good to know. So far, I've only tested nglide with 1964 but came across the same issue. I noticed Alt+Enter worked better. So i guess the full screen issue isn't the emulator's fault.

Do you think it is possible to play it in windowed mode? I don't get why it requires full screen :disappointed: . I actually prefer to play in window mode, but nglide + Glide64 final is still my HLE gfx plugin of choice :smile: .

**Comment by project64** [comment]
maybe I should focus on trying angle once I get 2.3 released, would be nice to make a d3d version of glide64 default.

**Comment by Frank-74** [comment]
The first time I ran it with nGlide, Project 64 forced it to window mode after I pressed Escape to go fullscreen.  I can't work out why it worked in window mode the first time I ran it, but not since.  Might have been the resolution I had it set too.  I run it at the max fullscreen resolution I can now, with no slowdown compared to 640x480.

**Comment by cxd4** [comment]
I can just see the walls of GitHub emoji by Wally123 already.

**Comment by Frank-74** [comment]
@LegendOfDragoon 
I can use it in window mode now.

I set screen resolution in nGlide configurator to 800x600, aspect ratio to preserve original.  Glide I set screen to 800x600 also.

Press Alt and tap Enter, It will probably go fullscreen first, Alt+Enter again it will go window mode.  But without the PJ64 status bar.

I still have to end with F12 though.

**Comment by LegendOfDragoon** [comment]
Yeah, i noticed that too actually. Seems kinda buggy still. Definitely worth using though. I sadly wanted to view the status bar too, so I could see VI/s among other things. Good thing that the difference in performance is so big, I don't need to see the numbers :smile: .

>maybe I should focus on trying angle once I get 2.3 released, would be nice to make a d3d version of glide64 default.

That would be awesome! It gives me incentive to work on HLE instead of LLE :smile: .

**Comment by LegendOfDragoon** [comment]
@Frank-74 have you tried testing frame buffer emulation with NGlide? I get much better performance in Mario Kart :smile: .

However, I don't think it's worth the effort to get Nglide to work on PJ64's Glide64 plugin, because it's closed source. I don't mind just using Glide64 Final for the time being. I think it would be more worthwhile to try out Angle.

**Comment by Frank-74** [comment]
I have tested and do get better performance.

However, some effects in games make Glide64 choke.  In Perfect Dark, using nGlide, when the enemy use an invisible cloak, nGlide doesn't handle it, but PJGlide does.

**Comment by Frank-74** [comment]
> You think the problem is the wrapper? Or the fact that PJ64's Glide has improvements?

Problem is the wrapper, Glide64 Final with the normal wrapper works ok.

**Comment by cxd4** [comment]
And the GL in Glide64's normal wrapper, Glitch64, is halfway still coded with slow, obsolete techniques.

**Comment by LegendOfDragoon** [comment]
>Problem is the wrapper, Glide64 Final with the normal wrapper works ok.
Also, 2d menu's in Monaco Grand Prix with nGlide drag at 1-2 VI's on fade in/out. PJGlide and Final's normal wrapper is ok.

I see. I did notice that Bangaioh didn't look the same when using NGlide, but it's buggy w/o Nglide anyway so I have no reason to even use Glide64 for it lol..

>And the GL in Glide64's normal wrapper, Glitch64, is halfway still coded with slow, obsolete techniques.

Indeed. Even z64gl has faster framebuffer code.

**Comment by cxd4** [comment]
Direct3D is documented a lot better and a lot more professionally by affiliations with Microsoft, though, so it really wouldn't surprise me if nGlide was just coded better.  This is what I mean about nGlide not necessarily being a solution--it doesn't mean one API is faster than the other; it just indicates one wrapper performs better than the other.

As far as Google ANGLE goes though, it's a wrapper written in Direct3D which can probably be configured or replaced with one of ATI's or NVIDIA's OpenGL wrappers for the desktop port of zilmar's Android port of...the desktop port of Glide, or whatever the heck he's doing to it.

**Comment by LegendOfDragoon** [comment]
I certainly think that nGlide is just more optimized. I couldn't even get similar fbe performance with Rice, without resorting to things like using SSE. I know OpenGL can be fast, thanks to Rice :smile: . His OpenGL plugin is faster than Jabo's. So I don't blame serious performance issues on APIs.

My main incentive in using D3D would be drivers.  If I had/knew a good profiler for measuring OpenGL performance, I'd be more inclined to just profile Glitch64 and see if I could improve it.

**Comment by project64** [comment]
I am looking forward to trying to get ANGLE working and see if there is any performance difference. I did notice that on linux it can use opengl so that is all good there. Just need to get the android version to a place where I can release a public beta first.

**Comment by cxd4** [comment]
I'm not sure but I'm guessing ANGLE could probably use opengl on Windows as well.  All browsers supporting WebGL, except for Internet Explorer 11 (I think Edge too), not only default to using Direct3D to wrap the WebGL opengl ES calls but also are configurable to disable ANGLE and not use Direct3D.

>My main incentive in using D3D would be drivers.

Yep pretty sure there are more drivers for Windows where Direct3D is implemented faster than OpenGL is.  With newer drivers however (particularly with NVIDIA) this becomes less true overtime.  With Intel it used to be that both GL and Direct3D sucked for n64 emulation; this stopped being reputable around 2008 or so.  Now it's only D3D that doesn't suck with Intel. :) There's this thing called money so that's why.

Anything DirectX under COM is cleaner when written as C++ though, not C.  Doing it in C goes through some macros for the native C++ binding for things like Direct3D, which I find annoying.  GL is equally compatible with both C and C++, and you can also query for extensions beyond the version of the GL context you initialized, whereas with Direct3D you have a fixed definitive range of features per the version of Direct3D number you initialized.

An ideal application could query the card for futuristic, very fast techniques and extensions in OpenGL and, if not found, fall back to Direct3D on Windows for better performance.  Most of these ARB extensions to OpenGL are implemented in Direct3D secondhand as the APIs become less and less different from each other overtime in terms of supported feature sets.

>If I had/knew a good profiler for measuring OpenGL performance, I'd be more inclined to just profile Glitch64 and see if I could improve it.

It's easier to notice performance differences when writing test applications and renderings than just testing both inside of a N64 emulator I think.  Also, WebGL (with ANGLE optionally enabled/disabled for Direct3D or OpenGL emulation of ES) opengl commands can be profiled and benchmarked through Internet Explorer's JavaScript debugging console.  It has a very good profiler feature for called JavaScript functions, including WebGL commands.  FireFox's is alright as well.

**Comment by LegendOfDragoon** [comment]
>Probably nGlide's framebuffer emulation is fast because it's incomplete.

What makes you say that? I know Mario Kart's jumbotron isn't particularly big, but I doubt nglide takes advantage of partial framebuffer writes. I'll have to look at Glide64's code to verify though.

>An ideal application could query the card for futuristic, very fast techniques and extensions in OpenGL and, if not found, fall back to Direct3D on Windows for better performance.

I think that's a neat idea. It's great to have some sort of fallback for certain users.

>It's easier to notice performance differences when writing test applications and renderings than just testing both inside of a N64 emulator I think.

Idk why I didn't think of this :smile: . I guess you're right. I'll probably do that for figuring out how to optimize blitting and frame buffer reads & writes. I used to do this a lot for testing x86 algorithms, but I eventually gravitated towards real time performance profiling. Maybe I'll look into testing WebGL someday.

**Comment by cxd4** [comment]
As server-side graphics code in DirectX/OpenGL run in a different dimension and thread than the C program for the CPU flow itself, profiling and benchmarking GL/DX is something of another philosophy.

Due to this, it really wouldn't surprise me if there are other options besides what I mentioned, but I never really cared to ask the question.  It seems to miss the point.  With C code you have multiple aspects:  machine code generated from compiler optimizations, different processors and different latencies and throughputs causing different profiler results with different algorithms.

Contrarily, with OpenGL code, general wisdom and education seems, by comparison, to more likely outweigh the scientific model of experimentally profiling the program for microperformance numbers.  The closest thing to profilable C code level in DX or GL is shader scripting in HLSL or GLSL (respectively); everything else is just high-level functions like `glDrawArrays` that lose meaning for microperformance profiling under the analogous, high-level reasoning of when, where and why to invoke it.

**Comment by LegendOfDragoon** [comment]
What version of OGL is the current Glide 64 wrapper? I'm interested in finally learning OGL :smile:

**Comment by cxd4** [comment]
I don't recommend thinking of extensions-based APIs in terms of version numbers.

**Comment by project64** [comment]
I started to play with angle (https://github.com/project64/project64/tree/angle) tho it is not working at the moment, it feels like it is close to working

**Comment by LegendOfDragoon** [comment]
>I don't recommend thinking of extensions-based APIs in terms of version numbers.

Good point :smile: . I'll keep that in mind. I'm basically trying to decide where to set the limit, as far as what extensions to use. I'm hoping Glide64 doesn't need anything past OGL 1.5. My goal is to have decent quality emulation with low system requirements.

>I started to play with angle (https://github.com/project64/project64/tree/angle) tho it is not working at the moment, it feels like it is close to working

Sweet, I look forward to seeing how well it runs, once it works. Guess I'll put time into learning Angel in the near future.

**Comment by cxd4** [comment]
Vertex and fragment shader assembly using the programmable pipeline requires extensions from at least the OpenGL 2.0 specifications.  AFAIR 1.5 introduces some optimizations for DMA to video memory; compared to the rest of improvements with a 3-D hardware-accelerated renderer plugin I think that this is not that significant relatively speaking.

You're always free to implement GL2 or GL 3.x rendering of the command lists, then re-assign the corresponding function pointers to alternative functions using GL 1.1 to GL 1.5 code for systems that lack those extensions.

Unlike as with DirectX, this means it's possible that a plugin can use OpenGL 4.4 on a system that supports it but fall back to OpenGL 1.1 in theory if you had all that organization laid out.

**Comment by LegendOfDragoon** [comment]
I guess I'll mainly focus on learning OGL 2.0 extensions then. For some reason I thought that 1.5 could do enough, but perhaps not.

I do like the idea of being able to support alternative functions :smile: .

**Comment by cxd4** [comment]
The extensions hardly have much to do with version numbers.  Saying that there's a GL 2.0 extension or a GL 3.2 extension only indicates the time frame where the ARB unanimously recognized the feature as widely useful and implemented.

With COM or DirectX a version number indicates that you must have X minimum version to use Y feature, just because it was implemented in the interface that late.  With OpenGL it's just a version number and little else.

The only thing that having OpenGL 2.0 contexts does is regulate that the function must exist.  If it's an extension running in GL 1.5 or earlier then it hardly matters, because your drivers have this feature implemented because they chose to, not because they were forced to come up with something hacky just so they could brand it as a 2.0 chip.
